pkg/lease/manager.go

```diff
@@ -0,0 +1,205 @@
+
+}
+
+func NewManagerWithCustomLogger(cl client.Client, log logr.Logger) Manager {
+	return &manager{
+		Client: cl,
+		log:    log,
+	}
+}
+
+func (l *manager) createOrGetExistingLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, leaseNamespace string) (*coordv1.Lease, bool, error) {
+	owner := makeExpectedOwnerOfLease(node)
+	microTimeNow := metav1.NowMicro()
+
+	lease := &coordv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
+			Namespace:       leaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: pointer.Int32(int32(duration.Seconds())),
+			AcquireTime:          &microTimeNow,
+			RenewTime:            &microTimeNow,
+			LeaseTransitions:     pointer.Int32(0),
+		},
+	}
+
+	if err := l.Client.Create(ctx, lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+			nodeName := node.ObjectMeta.Name
+			key := apitypes.NamespacedName{Namespace: leaseNamespace, Name: nodeName}
+
+			if err := l.Client.Get(ctx, key, lease); err != nil {
+				return nil, false, err
+			}
+			return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false, nil
+}
+
+func (l *manager) updateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error) {
+	needUpdateLease := false
+	setAcquireAndLeaseTransitions := false
+	updateAlreadyOwnedLease := false
+
+	if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity == holderIdentity {
+		needUpdateLease, setAcquireAndLeaseTransitions = needUpdateOwnedLease(lease, *currentTime, leaseDeadline)
+		if needUpdateLease {
+			updateAlreadyOwnedLease = true
+
+			log.Infof("renew lease owned by nmo setAcquireTime=%t", setAcquireAndLeaseTransitions)
+
+		}
+	} else {
+		// can't update the lease if it is currently valid.
+		if isValidLease(lease, currentTime.Time) {
+			return false, fmt.Errorf("can't update valid lease held by different owner")
+		}
+		needUpdateLease = true
+
+		log.Info("taking over foreign lease")
+		setAcquireAndLeaseTransitions = true
+	}
+
+	if needUpdateLease {
+		if setAcquireAndLeaseTransitions {
+			lease.Spec.AcquireTime = currentTime
+			if lease.Spec.LeaseTransitions != nil {
+				*lease.Spec.LeaseTransitions += int32(1)
+			} else {
+				lease.Spec.LeaseTransitions = pointer.Int32(1)
+			}
+		}
+		owner := makeExpectedOwnerOfLease(node)
+		lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+		lease.Spec.HolderIdentity = &holderIdentity
+		lease.Spec.LeaseDurationSeconds = pointer.Int32(int32(leaseDuration.Seconds()))
+		lease.Spec.RenewTime = currentTime
+		if err := l.Client.Update(ctx, lease); err != nil {
+			log.Errorf("Failed to update the lease. node %s error: %v", node.Name, err)
+			return updateAlreadyOwnedLease, err
```

Comment:
the meaning / use case of the 1st return value is mysterious to me :D

Reply:
I always question code from the past, no matter who wrote it, including me 🙂 

> whether the update was successful or not

that's what we have the error for, not? 🤔

Reply:
Looking into this further I think that the bool is an indication whether prior to the update the lease was already held by holderIdentity (true) or not (false).